### ISSUE TITLE
Nuanced rules when santising? (to enable other attributes like target + rel)

### ISSUE THREAD

**Issue by Zeouterlimits** [rails/rails-html-sanitizer#107]
For the allowed attributes in html:

https://github.com/rails/rails-html-sanitizer/blob/93eb2491daa0f758e1e4d3e72833c911dfb71468/lib/rails/html/sanitizer.rb#L111

Would it be possible to make the rules more nuanced? 
For example, the target attribute is not included by default, likely because it can be used to exploit a browsers window.opener api.
However that only applies, when the rel attribute isn't present, with the right inputs

Explanation:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#target

So it would be cool if target could be a valid attribute, when present with the correct rel attribute?

I'll take a look at implementing it, but points welcome :)

**Comment by flavorjones** [comment]
@Zeouterlimits I know it's been a while since you opened this, but can you help me understand your use case? I'm curious when you'd want to allow this specific attribute through in certain circumstances in untrusted input.

The scope of `rails-html-sanitizer` is for making _untrusted_ input safe, and it's not really designed for nuanced conditional handling like you're describing. But if you tell me more about your use case, maybe I can suggest an alternative solution.

**Comment by dankimio** [comment]
@flavorjones I have a simialar issue where I want to extend the existing `SafeListSanitizer` to allow links with `target="_blank"`, but I couldn't find an easy way to do it.

**Comment by dankimio** [comment]
> @flavorjones I have a simialar issue where I want to extend the existing `SafeListSanitizer` to allow links with `target="_blank"`, but I couldn't find an easy way to do it.

Came up with this:

```ruby
Rails::Html::SafeListSanitizer.allowed_attributes << "target"
```

**Comment by QuantumSand41Fox** [comment]
I use the following code to allow target blank in link tags using `sanitize`.

config/initializers/sanitizers.rb:
```ruby
require 'loofah'

# override view layer's sanitizer to support <a target="_blank"></a> (used e.g. in rich text editor)
# (also control layer's ActionController::Base.helpers.sanitize is overriden)
module ActionView
    module Helpers
      module SanitizeHelper
        alias_method :original_sanitize, :sanitize

        def sanitize(html, options = {})
          doc = Loofah.fragment(html)
          doc.scrub!(Loofah::Scrubber.new { |node|
              if node.name == 'a' && node['target'] == '_blank'
                node['rel'] = 'noopener noreferrer'
              end
          })

          return original_sanitize(doc.to_s, options)
        end
      end
    end
  end
```

 config/appication.rb:
```ruby
config.after_initialize do
	# Allow target and rel attributes. Safety is handled in /config/initializers/sanitizers.rb
	ActionView::Base.sanitized_allowed_attributes.add 'target'
	ActionView::Base.sanitized_allowed_attributes.add 'rel'
end
```

**Comment by flavorjones** [comment]
@QuantumSand41Fox I'm currently going through some design ideas for improving the flexibility of the Rails sanitizer, so if you don't mind I'd love to ask some questions ...

In the case where you need to allow `target="_blank"`, where is the data coming from? Is this untrusted content, or is it markdown that you've written in your app that happens to be passing through the sanitizer?

Thanks for any additional context you can provide about your use case.

**Comment by QuantumSand41Fox** [comment]
@flavorjones 
Of course. In my case the data is untrusted and it is coming form a rich text editor called Tinymce, where an user can specify that a link opens in a new tab.

**Comment by flavorjones** [comment]
@QuantumSand41Fox Thank you for sharing the details! That's a good use case that I will keep in mind if and when we make Rails::HTML::Sanitizer a bit more extensible.
